### ISSUE TITLE
False positive [attr-defined] when using namespace_packages

### ISSUE THREAD

**Issue by dbdbc** [python/mypy#13663]
**Bug Report**

When using `namespace_packages = true`, if it happens that a directory with the same name as a module exists, `mypy` will not recognize names from the module.

**To Reproduce**

This is a minimal example, tested with python 3.8 and mypy 0.971:
```
.
├── mypkg
│   ├── foo  # empty directory
│   ├── foo.py
│   └── bar.py
└── mypy.ini
```

```
# mypy.ini
[mypy]
namespace_packages = true
show_error_codes = true
```

```
# mypkg/bar.py
from mypkg.foo import func
```

```
# mypkg/foo.py
def func():
    pass
```

While python correctly ignores the empty directory `mypkg/foo`, mypy does not:
```bash
# All these are successful
$ python -c 'import mypkg.bar'
$ python -c 'import mypkg.foo'
$ python -c 'from mypkg.foo import func'

# These calls to mypy fail
$ mypy -p mypkg
mypkg/bar.py:1: error: Module "mypkg.foo" has no attribute "func"  [attr-defined]
Found 1 error in 1 file (checked 3 source files)
$ mypy mypkg/bar.py 
mypkg/bar.py:1: error: Module "mypkg.foo" has no attribute "func"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```

**Expected Behavior**

Mypy should ignore the empty directory and run successfully on this package. According to [the docs](https://mypy.readthedocs.io/en/stable/running_mypy.html#how-imports-are-found), a package should be found (and have precedence over a module) if a directory has `__init__.py` or `__init__.pyi` in it.

In the case of `namespace_packages = true`, it seems that mypy is satisfied with a directory missing an `__init__.py` file and does not look for the module of the same name. This differs from the python search order specified in [PEP-420](https://peps.python.org/pep-0420/#specification), where the following are searched in turn: `<directory>/foo/__init__.py`, `<directory>/foo.{py,pyc,so,pyd}`, `<directory>/foo`, meaning that a namespace package is the last priority.

**Actual Behavior**

Mypy detects `mypkg/foo` as a package before module `mypkg/foo.py` and does not recognize the names in `mypkg/foo.py`, resulting in false positive `[attr-defined]` errors.

**Your Environment**

- Mypy version used: 0.971
- Mypy command-line flags: `-p mypkg` or `mypkg/bar.py`
- Mypy configuration options from `mypy.ini` (and other config files): `namespace_packages = true`
- Python version used: 3.8.5
- Operating system and version: Linux (Ubuntu 18.04)

**Comment by dbari** [comment]
FYI I'm the original reporter (had to merge my GitHub accounts)

**Comment by prpawar** [comment]
any update on this ?

**Comment by zwimer** [comment]
I've run into a similar issue before. I hope this gets addressed
